### PR TITLE
Parse function improvements

### PR DESCRIPTION
**Pull request by kossnocorp**
- Stand-alone era token now parses as `1 BC` and `1 AD` instead of `10 BC` and `10 AD`.
-  Fix bug when week year was used to parse Gregorian year and the Gregorian year was used to parse week year.
- Validate parse tokens (see #1220)
- Rename baseDate to backupDate in parse (see #1219)
- Remove locale-independent tokens from locale snapshots (see #1221)

Also, @babel/plugin-transform-parameters and @babel/plugin-transform-destructuring were added to the Babel config.

### CODE REVIEW

File: src/parse/index.js [hunk 1]
```diff
@@ -454,6 +466,25 @@ export default function parse(
     var firstCharacter = token[0]
     var parser = parsers[firstCharacter]
     if (parser) {
+      const { incompatibleTokens } = parser
```
Comment (leshakoss) [inline src/parse/index.js]: Are we using `const` now?

File: src/parse/index.js [hunk 2]
```diff
@@ -454,6 +466,25 @@ export default function parse(
...
+      if (Array.isArray(incompatibleTokens)) {
+        const incompatibleToken = usedTokens.find(
```
Comment (leshakoss) [inline src/parse/index.js]: `find` is not available for IE, we have to rewrite it with `for` or provide a polyfill

File: src/parse/index.js [hunk 3]
```diff
@@ -454,6 +466,25 @@ export default function parse(
...
+          ({ token }) =>
+            incompatibleTokens.indexOf(token) !== -1 || token === firstCharacter
```
Comment (leshakoss) [inline src/parse/index.js]: Same for `indexOf`, only available since IE9

File: src/parse/_lib/parsers/index.js [hunk 4]
```diff
@@ -282,12 +283,16 @@ var parsers = {
           )
       }
     },
-    set: function(date, _flags, value, _options) {
+
+    set: function(date, flags, value, _options) {
       // Sets year 10 BC if BC, or 10 AC if AC
-      date.setUTCFullYear(value === 1 ? 10 : -9, 0, 1)
+      flags.era = value
+      date.setUTCFullYear(value === 1 ? 1 : 0, 0, 1)
```
Comment (leshakoss) [inline src/parse/_lib/parsers/index.js]: A nitpick: interestingly value here could only be 0 or 1 so the statement could be simplified to `date.setUTCFullYear(value, 0, 1)`

File: src/parse/index.js (hunk 1 above)
Comment (kossnocorp) [reply src/parse/index.js]: Yup, several times I've shipped `const`, template literals and arrow functions so I handpicked Babel plugins to transpile those: https://github.com/date-fns/date-fns/blob/master/.babelrc.js#L2-L6

Also, I've added a few in the PR: https://github.com/date-fns/date-fns/pull/1224/commits/e244d27d76a954cdc899f85e9932b0032c6686c7

File: src/parse/_lib/parsers/index.js (hunk 4 above)
Comment (kossnocorp) [reply src/parse/_lib/parsers/index.js]: Wow, I didn't realize that. Good catch!

File: src/parse/index.js (hunk 2 above)
Comment (kossnocorp) [reply src/parse/index.js]: God damn, IE!

File: src/parse/index.js (hunk 3 above)
Comment (kossnocorp) [reply src/parse/index.js]: Let's raise the bar up to IE11.

File: src/parse/_lib/parsers/index.js [hunk 5]
```diff
@@ -282,12 +283,16 @@ var parsers = {
           )
       }
     },
-    set: function(date, _flags, value, _options) {
+
+    set: function(date, flags, value, _options) {
       // Sets year 10 BC if BC, or 10 AC if AC
```
Comment (leshakoss) [inline src/parse/_lib/parsers/index.js]: That comment is outdated :)

Comment (kossnocorp) [reply src/parse/_lib/parsers/index.js]: Right!

**Pull request merged**